### ISSUE TITLE
Remodel derivation mechanism

### ISSUE THREAD

**Issue by sebasguts** [homalg-project/CAP_project#144]
Since several issued with the derivation mechanism popped up recently, there should be a complete remodeling of the derivation mechanism. The following topics should be adressed:

* Derivations should only happen when the category is finalized, not on the fly. A category is not guaranteed to function if not finalized anyway, so this is not a breaking change in any sense. Nevertheless, it is a change in the behaviour of categories and might need to different derivations triggered then now.

* The previous point can lead us to getting rid of the final derivations all together. Each derivation should now possibly contain a set of exclusions, in addition to a set of requirements. This leads to the fact that all derivations will be like the current final derivations.

* Derivations should be able to dependend on specific other derivations, not only on primitives. Some derivations make assumptions about how other primitives are derived. Those have to be installed in a group. It might not be flexible enough to actually group them, since some derivations in a group might work without the others in the group, and others wont.

To adress those changes, the following changes in the CAP kernel are necessary:

* [ ] Change the API for the derivations to be compatible with the one for final derivations, and remove final derivations.
* [ ] Change the method derivation graph to be able to work with the new requirements
* [ ] Change the algorithm that computes the derivations to compute a maximal tree in the derivation graph, where the term *maximal* might be still need discussion.

This issue is meant to track progress on this topic, and will be updated in the next few days, while I am working on implementing this feature.

**Comment by zickgraf** [comment]
Once https://github.com/homalg-project/CAP_project/pull/1699 is merged, I think this issue can be closed, after more than 6 years!

Changes implemented over the last years:
* Derivations only happen when the category is finalized.
* Final derivations can be grouped into bundles.
* The mechanisms for final and for ordinary derivations now both reside in `Derivations.gi` and share a lot of code.

Changes not implemented:
* Final and ordinary derivations are still different concepts.
* Final derivations are applied blindly in the order in which they are added to CAP.

In particular, there is no mechanism for finding a *maximal* tree of derivations. However, such a mechanism would come with its own drawbacks:
* The current mechanism is easy to explain: First apply as many ordinary derivations as possible, afterwards try to apply final derivations. A mechanism finding a *maximal* tree of derivations would be much more difficult to explain.
* The current mechanism seems to produce stable results in the sense that adding new derivations does usually not affect whether existing derivations are triggered. When finding a *maximal* tree of derivations however, I can imagine a situation where adding new derivations (e.g. by loading another package) suddenly leads to completely different derivations being triggered.

Summing up, I think the current state of the derivation mechanism is future-proof and covers the original problems given in this issue well enough to close this issue as completed.

**Comment by mohamed-barakat** [comment]
Wonderful, thank you very much.
